### PR TITLE
Ignore extra COLLATE default which can come as part of default value of column and ignore explicit COLLATE default on view def in dump

### PR DESCRIPTION
**Pull request by Deepesh125**
### Description

In upgraded database, from Babelfish v2.1.0/v2.2.0 to Babelfish v2.3.0, 
1. we might see extra COLLATE "default" for default value clause of table column in dump.
2. we might see extra COLLATE "default" around const clause in view definition which may lead collation conflict error during restore and hence may lead to upgrade failure.
3. We might also see extra COLLATE "default" clause around check constraints. 

This commit takes care of extra COLLATE clause for above mentioned cases. Also, this commit removes temporary workaround made for _ci_sysname data type to pick correct collation. 

Note: This is temporary fix to support major version upgrade from 14.6 to 15.1. Correct solution will be made to 14.6.1 onwards. 
 
Extension test PR: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/945
Task: BABEL-3801, BABEL-3814
Signed-off-by: Dipesh Dhameliya <dddhamel@amazon.com>
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/utils/adt/ruleutils.c [hunk 1]
```diff
@@ -10549,6 +10549,20 @@ get_const_collation(Const *constval, deparse_context *context)
...
+		 * So in upgraded instances to 2.3.0 from 2.2.0, constval->constcollid would be DEFAULT_COLLATION_OID
+		 * if object was created prior to 2.3.0 and typcollation would be not equal to DEFAULT_COLLATION_OID.
+		 * So avoid dumping COLLATE cluase in such case.
+		 */
+		if (babelfish_dump_restore && 
+			 strncmp(babelfish_dump_restore, "on", 2) == 0 &&
+			 constval->constcollid ==  DEFAULT_COLLATION_OID && 
+			 typcollation != DEFAULT_COLLATION_OID &&
+			 constval->constcollid != typcollation)
+			return;
```
Comment (Deepesh125) [inline src/backend/utils/adt/ruleutils.c]: This is unnecessary change, will remove it in final rev.

File: src/backend/parser/gram.y [hunk 2]
```diff
@@ -18612,9 +18612,34 @@ SplitColQualList(List *qualList,
 		}
 		if (IsA(n, CollateClause))
 		{
+			const char *dump_restore = GetConfigOption("babelfishpg_tsql.dump_restore", true, false);
 			CollateClause *c = (CollateClause *) n;
 
-			if (*collClause)
+			/*
+			 * XXX: This is temporary fix.
+			 * During upgrade from 14.6 to 15.1, we can ignore extra COLLATE "default" clause.
```
Comment (rishabhtanwar29) [inline src/backend/parser/gram.y]: We should not mention specific engine versions since we don't know in which release this will get fixed.

File: src/backend/parser/parse_collate.c [hunk 3]
```diff
@@ -862,6 +863,28 @@ merge_collation_state(Oid collation,
...
 				if (collation != context->collation)
 				{
+					const char *dump_restore = GetConfigOption("babelfishpg_tsql.dump_restore", true, false);
+
+					/*
+					 * XXX: This is temporary fix. 
+					 * During upgrade from 14.6 to 15.1, we can ignore COLLATE default if
+					 * Babelfish database is being restored.
+					 */
+					if (dump_restore && strncmp(dump_restore, "on", 2) ==0)
```
Comment (rishabhtanwar29) [inline src/backend/parser/parse_collate.c]: Nit: missing whitespace after `==`

File: src/backend/parser/parse_collate.c [hunk 4]
```diff
@@ -862,6 +863,28 @@ merge_collation_state(Oid collation,
 			case COLLATE_EXPLICIT:
 				if (collation != context->collation)
 				{
+					const char *dump_restore = GetConfigOption("babelfishpg_tsql.dump_restore", true, false);
```
Comment (surendravishnoi) [inline src/backend/parser/parse_collate.c]: As we have seen, GetConfigOption could a perf issue. If there is flag to know if we are doing restore, we should put this code inside if (restore) condition

File: src/bin/pg_dump/dump_babel_utils.c [hunk 5]
```diff
@@ -484,3 +550,20 @@ dumpBabelfishSpecificConfig(Archive *AH, const char *dbname, PQExpBuffer outbuf)
 		pfree(current_server_collation_name);
 	}
 }
+
+/*
+ * babelfish_handle_view_def - Any upgraded Babelfish database from 2.1.0/2.2.0 to 2.3.0
+ * may dump extra COLLATE "default" clause around const clause inside view definition which
+ * is kind of unnecessary for tsql objects as well as for sys views. So this function remove
```
Comment (rishabhtanwar29) [inline src/bin/pg_dump/dump_babel_utils.c]: Nit: *removes

File: src/bin/pg_dump/dump_babel_utils.c [hunk 6]
```diff
@@ -139,24 +165,64 @@ bbf_selectDumpableCast(CastInfo *cast)
...
+	if (adef_expr != attrDefInfo->adef_expr)
+	{
+		free(attrDefInfo->adef_expr);
+		attrDefInfo->adef_expr = adef_expr;
+	}
+}
+
+/*
+ * fixTsqlCheckConstraint - In upgraded Babelfish database from v2.1.0/2.2.0 to v2.3.0 may contain extra
+ * COLLATE "default" clause around check constraint which is kind of redundant in tsql objects.
```
Comment (rishabhtanwar29) [inline src/bin/pg_dump/dump_babel_utils.c]: tsql -> T-SQL

**Review by rishabhtanwar29 — DISMISSED** [review]

File: src/bin/pg_dump/dump_babel_utils.c (hunk 5 above)
Comment (Deepesh125) [reply src/bin/pg_dump/dump_babel_utils.c]: Resolved!

File: src/bin/pg_dump/dump_babel_utils.c (hunk 6 above)
Comment (Deepesh125) [reply src/bin/pg_dump/dump_babel_utils.c]: Done!

File: src/bin/pg_dump/dump_babel_utils.c [hunk 7]
```diff
@@ -124,6 +124,32 @@ bbf_selectDumpableCast(CastInfo *cast)
 		cast->dobj.dump = DUMP_COMPONENT_NONE;
 }
 
+static char *
+removeSubstr(char *src, const char *substr)
```
Comment (surendravishnoi) [inline src/bin/pg_dump/dump_babel_utils.c]: Use bbf in function name

File: src/bin/pg_dump/dump_babel_utils.c [hunk 8]
```diff
@@ -124,6 +124,32 @@ bbf_selectDumpableCast(CastInfo *cast)
...
+	char *result = "";
+
+	if (!strstr(src, substr))
+		return src;
+
+	while ((str = strstr(src, substr)) != NULL)
+	{
+		result = psprintf("%s%s", result, pnstrdup(src, str - src));
+
+		/* skip "COLLATE \"default\" */
```
Comment (surendravishnoi) [inline src/bin/pg_dump/dump_babel_utils.c]: this is generic function, only say skip substr pattern

File: src/bin/pg_dump/dump_babel_utils.c [hunk 9]
```diff
@@ -139,24 +165,64 @@ bbf_selectDumpableCast(CastInfo *cast)
...
-		!strstr(source, runtimeErrStr) ||
-		strstr(source, runtimeErrFunc) ||
-		attrDefInfo->adnum < 1)
+	if (!isBabelfishDatabase(fout) || attrDefInfo->adnum < 1)
 		return;
 
+	/* Handle default value for "numeric" and "decimal") */
 	atttypname = attrDefInfo->adtable->atttypnames[attrDefInfo->adnum - 1];
-	if (!strstr(atttypname, "decimal") && !strstr(atttypname, "numeric"))
+	if (strstr(atttypname, "decimal") || strstr(atttypname, "numeric"))
```
Comment (surendravishnoi) [inline src/bin/pg_dump/dump_babel_utils.c:190]: have we made sure that collate default is not applicable to decimal and numeric?

File: src/bin/pg_dump/dump_babel_utils.c [hunk 10]
```diff
@@ -124,6 +124,32 @@ bbf_selectDumpableCast(CastInfo *cast)
...
+{
+	char *str;
+	char *result = "";
+
+	if (!strstr(src, substr))
+		return src;
+
+	while ((str = strstr(src, substr)) != NULL)
+	{
+		result = psprintf("%s%s", result, pnstrdup(src, str - src));
```
Comment (surendravishnoi) [inline src/bin/pg_dump/dump_babel_utils.c]: result is getting leaked in the loop except for last time

File: src/bin/pg_dump/dump_babel_utils.c [hunk 11]
```diff
@@ -124,6 +124,32 @@ bbf_selectDumpableCast(CastInfo *cast)
...
+
+		/* skip "COLLATE \"default\" */
+		src = str + strlen(substr);
```
Comment (surendravishnoi) [inline src/bin/pg_dump/dump_babel_utils.c]: strlen can be calc outside the loop

File: src/bin/pg_dump/dump_babel_utils.c [hunk 12]
```diff
@@ -484,3 +550,20 @@ dumpBabelfishSpecificConfig(Archive *AH, const char *dbname, PQExpBuffer outbuf)
...
+ * may dump extra COLLATE "default" clause around const clause inside view definition which
+ * is kind of unnecessary for T-SQL objects as well as for sys views. So this function removes 
+ * such clause from view definition. 
+ */
+char *
+babelfish_handle_view_def(Archive *fout, char *view_def)
+{
+	const char *defaultCollation = "COLLATE \"default\"";
+
+	if (!isBabelfishDatabase(fout))
```
Comment (surendravishnoi) [inline src/bin/pg_dump/dump_babel_utils.c:578]: can we add the check if it is a TSQL view?

File: src/bin/pg_dump/dump_babel_utils.c (hunk 7 above)
Comment (Deepesh125) [reply src/bin/pg_dump/dump_babel_utils.c]: Done

File: src/bin/pg_dump/dump_babel_utils.c (hunk 8 above)
Comment (Deepesh125) [reply src/bin/pg_dump/dump_babel_utils.c]: Yes, Updated in new rev.

File: src/bin/pg_dump/dump_babel_utils.c (hunk 11 above)
Comment (Deepesh125) [reply src/bin/pg_dump/dump_babel_utils.c]: Done!

File: src/bin/pg_dump/dump_babel_utils.c (hunk 9 above)
Comment (Deepesh125) [reply src/bin/pg_dump/dump_babel_utils.c:190]: It is not.

File: src/bin/pg_dump/dump_babel_utils.c (hunk 10 above)
Comment (Deepesh125) [reply src/bin/pg_dump/dump_babel_utils.c]: Handled in latest rev.

**Review by surendravishnoi — APPROVED** [review]
